care/emr/api/viewsets/meta_artifact.py

```diff
@@ -0,0 +1,128 @@
+        if permission == "read":
+            allowed = AuthorizationController.call(
+                "can_view_clinical_data", user, patient_obj
+            )
+        elif permission == "write":
+            allowed = AuthorizationController.call(
+                "can_write_patient_obj", user, patient_obj
+            )
+    elif associating_type == MetaArtifactAssociatingTypeChoices.encounter.value:
+        encounter_obj = get_object_or_404(Encounter, external_id=associating_id)
+        if permission == "read":
+            allowed = AuthorizationController.call(
+                "can_view_clinical_data", user, encounter_obj.patient
+            ) or AuthorizationController.call(
+                "can_view_encounter_obj", user, encounter_obj
+            )
+
+        elif permission == "write":
+            allowed = AuthorizationController.call(
+                "can_update_encounter_obj", user, encounter_obj
+            )
+
+    if not allowed:
+        raise PermissionDenied("Cannot view object")
+
+
+class MetaArtifactViewSet(
+    EMRCreateMixin,
+    EMRRetrieveMixin,
+    EMRUpdateMixin,
+    EMRUpsertMixin,
+    EMRListMixin,
+    EMRBaseViewSet,
+):
+    database_model = MetaArtifact
+    pydantic_model = MetaArtifactCreateSpec
+    pydantic_read_model = MetaArtifactReadSpec
+    pydantic_update_model = MetaArtifactUpdateSpec
+    filterset_class = MetaArtifactFilters
+    filter_backends = (filters.DjangoFilterBackend,)
+
+    def authorize_create(self, instance):
+        meta_artifact_authorizer(
+            self.request.user,
+            instance.associating_type,
+            instance.associating_id,
+            "write",
+        )
+
+    def authorize_update(self, request_obj, model_instance):
+        meta_artifact_authorizer(
+            self.request.user,
+            model_instance.associating_type,
+            model_instance.associating_external_id,
+            "write",
+        )
+
+    def get_queryset(self):
+        if self.action == "list":
+            if (
+                "associating_type" not in self.request.GET
+                and "associating_id" not in self.request.GET
+            ):
+                raise PermissionDenied(
+                    "'associating_type' and 'associating_id' query params are required to list meta artifacts"
+                )
+            meta_artifact_authorizer(
+                self.request.user,
+                self.request.GET.get("associating_type"),
+                self.request.GET.get("associating_id"),
+                "read",
+            )
+            return (
+                super()
+                .get_queryset()
+                .filter(
+                    associating_type=self.request.GET.get("associating_type"),
+                    associating_external_id=self.request.GET.get("associating_id"),
+                )
+            )
+
+        obj = get_object_or_404(MetaArtifact, external_id=self.kwargs["external_id"])
```

Comment:
Can we move this to the top and change the logic so that we fetch the type and id based on what the method is and the rest of the code has a single flow ?

Reply:
done; i assume this was what's intended...